In [27]:
# Import datasets 
from sklearn.datasets import load_iris

In [28]:
# Loading data into an object
iris_data = load_iris()

Notice that the target are numeric values and not the names of the labels
The second command prints out the name

Reference:
- 0 : setosa
- 1 : versicolor
- 2 : virginica

In [29]:
# Quick visual on how the results are labeled in the dataset

print(iris_data.target) 
print(iris_data.target_names)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
['setosa' 'versicolor' 'virginica']


## Four implementation that scikit-learns expects

- Features and response are in separate objects
- Features and response must be numeric
- Features and response must be in NumPy arrays
- Features and response should have specific shapes

In [30]:
# Double check the shape of the features and response
print(iris_data.data.shape)
print(iris_data.target.shape)

(150, 4)
(150,)


Result: 
- The data has four features with 150 observations
- The target has one result with 150 results

In [31]:
# Store the data into variables
X_feat = iris_data.data 
y_trgt = iris_data.target

# Note: The X is uppercase because it represents a matrix while the y is lowercase because it's a vector

The KNN classification is an intersting method. You first decide how mnay points you want to choose, based on some technique. You then have your features without any class, or label. Thus, you look in the data set that most closely resembles with the data points for your unknown label. 


In [32]:
# Import the Kneighbors module
from sklearn.neighbors import KNeighborsClassifier

In [33]:
# The parameter n_neighbors=1 specifies that we need the ONE closest neighbor
knn = KNeighborsClassifier(n_neighbors=1)

In [34]:
# Fit the model with data
knn.fit(X_feat, y_trgt)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=1, p=2,
           weights='uniform')

In [35]:
# Quick example: Predicting the leaf based on some features
knn.predict([[2.5, 4, 5, 1.8]])


array([2])

In [36]:
# Predicting two observations
X_temp = [[3, 4, 4, 1], [2.5, 3.9, 4.1, 1.5]]

# Use the predict feature in k-neighbors
knn.predict(X_temp)

array([2, 2])

In [37]:
# Changing the paramet n_neighbors to 5. Instead of looking for the one closest neighbor, it'll look
# at the five closest and based off its prediction on those values
knn = KNeighborsClassifier(n_neighbors=5)

# Fit the model 
knn.fit(X_feat, y_trgt)

# Predict the X_temp with this modified model
knn.predict(X_temp)

array([1, 1])

In [38]:
# Using logistic regression
from sklearn.linear_model import LogisticRegression

# Instatiate the model
logreg = LogisticRegression()

# Fit model with values
logreg.fit(X_feat, y_trgt)

# Predict the X_temp with the logistic regression
log_pred = logreg.predict(X_temp)
log_pred # Results: They differ from the k-neighbors


array([2, 2])